// @partyglue/realtime/src/experiences/trivia/questions.ts
import type { TriviaQuestion } from "@partyglue/shared-types";

// Starter question bank — replace with DB-backed questions in Phase 3
// Categories: music (fits the DJ app theme), general, pop culture

export const SAMPLE_QUESTIONS: TriviaQuestion[] = [
  // ── MUSIC / DJ ──────────────────────────────────────────────────────────────
  {
    id: "q1",
    text: "What does BPM stand for in music?",
    options: [
      { id: "a", text: "Beats Per Minute" },
      { id: "b", text: "Bass Per Mix" },
      { id: "c", text: "Beats Per Measure" },
      { id: "d", text: "Base Pitch Mode" },
    ],
    correctOptionId: "a",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q2",
    text: "Which streaming platform launched in 2006 and pioneered music discovery?",
    options: [
      { id: "a", text: "Apple Music" },
      { id: "b", text: "Spotify" },
      { id: "c", text: "Tidal" },
      { id: "d", text: "SoundCloud" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q3",
    text: "What is the Camelot Wheel used for in DJing?",
    options: [
      { id: "a", text: "Tracking crowd energy" },
      { id: "b", text: "Harmonic key mixing" },
      { id: "c", text: "BPM synchronization" },
      { id: "d", text: "Stem separation" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q4",
    text: "Which artist holds the record for most weeks at #1 on Billboard Hot 100?",
    options: [
      { id: "a", text: "Drake" },
      { id: "b", text: "Taylor Swift" },
      { id: "c", text: "Mariah Carey" },
      { id: "d", text: "The Weeknd" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "medium",
  },
  {
    id: "q5",
    text: "What year did the first iPhone launch?",
    options: [
      { id: "a", text: "2005" },
      { id: "b", text: "2006" },
      { id: "c", text: "2007" },
      { id: "d", text: "2008" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q6",
    text: "What musical key is considered the 'saddest key' in classical music?",
    options: [
      { id: "a", text: "C Major" },
      { id: "b", text: "D minor" },
      { id: "c", text: "G Major" },
      { id: "d", text: "E flat Major" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "hard",
  },
  {
    id: "q7",
    text: "Which decade saw the rise of Hip-Hop as a mainstream genre?",
    options: [
      { id: "a", text: "1970s" },
      { id: "b", text: "1980s" },
      { id: "c", text: "1990s" },
      { id: "d", text: "2000s" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q8",
    text: "What is a 'drop' in EDM music?",
    options: [
      { id: "a", text: "The intro of a song" },
      { id: "b", text: "When the beat cuts out completely" },
      { id: "c", text: "The high-energy climax after a buildup" },
      { id: "d", text: "A DJ scratch technique" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q9",
    text: "Daft Punk is from which country?",
    options: [
      { id: "a", text: "Germany" },
      { id: "b", text: "USA" },
      { id: "c", text: "France" },
      { id: "d", text: "Sweden" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q10",
    text: "What does 'A&R' stand for in the music industry?",
    options: [
      { id: "a", text: "Audio & Recording" },
      { id: "b", text: "Artists & Repertoire" },
      { id: "c", text: "Albums & Releases" },
      { id: "d", text: "Acoustics & Resonance" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q11",
    text: "Which software is most popular among professional DJs for live mixing?",
    options: [
      { id: "a", text: "GarageBand" },
      { id: "b", text: "Ableton Live" },
      { id: "c", text: "Serato DJ" },
      { id: "d", text: "FL Studio" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q12",
    text: "How many semitones are in one octave?",
    options: [
      { id: "a", text: "8" },
      { id: "b", text: "10" },
      { id: "c", text: "12" },
      { id: "d", text: "16" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q13",
    text: "Which DJ is known as 'The Godfather of House Music'?",
    options: [
      { id: "a", text: "Tiësto" },
      { id: "b", text: "Frankie Knuckles" },
      { id: "c", text: "Carl Cox" },
      { id: "d", text: "Larry Heard" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "hard",
  },
  {
    id: "q14",
    text: "What does 'sampling' mean in music production?",
    options: [
      { id: "a", text: "Testing a track before release" },
      { id: "b", text: "Using a portion of an existing recording in a new track" },
      { id: "c", text: "Adding reverb to a vocal" },
      { id: "d", text: "Mastering audio at a lower volume" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q15",
    text: "What is 'sidechain compression' most commonly used for in EDM?",
    options: [
      { id: "a", text: "Tuning vocals" },
      { id: "b", text: "Making the bass pump in rhythm with the kick drum" },
      { id: "c", text: "Removing background noise" },
      { id: "d", text: "Speeding up the tempo" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "hard",
  },

  // ── POP CULTURE ─────────────────────────────────────────────────────────────
  {
    id: "q16",
    text: "Which artist released the album 'Lemonade' in 2016?",
    options: [
      { id: "a", text: "Rihanna" },
      { id: "b", text: "Cardi B" },
      { id: "c", text: "Beyoncé" },
      { id: "d", text: "Nicki Minaj" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q17",
    text: "What is the best-selling video game of all time?",
    options: [
      { id: "a", text: "Fortnite" },
      { id: "b", text: "Minecraft" },
      { id: "c", text: "Tetris" },
      { id: "d", text: "Grand Theft Auto V" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q18",
    text: "Which show features a 'Red Wedding'?",
    options: [
      { id: "a", text: "Westworld" },
      { id: "b", text: "The Witcher" },
      { id: "c", text: "Game of Thrones" },
      { id: "d", text: "Succession" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q19",
    text: "Drake is originally from which city?",
    options: [
      { id: "a", text: "New York" },
      { id: "b", text: "Atlanta" },
      { id: "c", text: "Toronto" },
      { id: "d", text: "Los Angeles" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q20",
    text: "What movie features the song 'All I Want for Christmas Is You'?",
    options: [
      { id: "a", text: "Elf" },
      { id: "b", text: "Home Alone" },
      { id: "c", text: "Love Actually" },
      { id: "d", text: "Jingle All the Way" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "pop culture",
    difficulty: "medium",
  },
  {
    id: "q21",
    text: "Which social media app introduced the 'Stories' format first?",
    options: [
      { id: "a", text: "Instagram" },
      { id: "b", text: "Facebook" },
      { id: "c", text: "Snapchat" },
      { id: "d", text: "TikTok" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "medium",
  },
  {
    id: "q22",
    text: "What year did TikTok launch internationally?",
    options: [
      { id: "a", text: "2016" },
      { id: "b", text: "2017" },
      { id: "c", text: "2018" },
      { id: "d", text: "2019" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "medium",
  },
  {
    id: "q23",
    text: "Which rapper's real name is Aubrey Graham?",
    options: [
      { id: "a", text: "Kendrick Lamar" },
      { id: "b", text: "Drake" },
      { id: "c", text: "J. Cole" },
      { id: "d", text: "Travis Scott" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q24",
    text: "What is the highest-grossing film of all time (not adjusted for inflation)?",
    options: [
      { id: "a", text: "Avengers: Endgame" },
      { id: "b", text: "Titanic" },
      { id: "c", text: "Avatar" },
      { id: "d", text: "Avatar: The Way of Water" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "pop culture",
    difficulty: "medium",
  },
  {
    id: "q25",
    text: "Billie Eilish recorded her first album in which unusual location?",
    options: [
      { id: "a", text: "A professional studio in LA" },
      { id: "b", text: "Her family's bedroom" },
      { id: "c", text: "A converted garage in London" },
      { id: "d", text: "A boat off the coast of California" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "pop culture",
    difficulty: "medium",
  },

  // ── GENERAL KNOWLEDGE ────────────────────────────────────────────────────────
  {
    id: "q26",
    text: "What is the capital of Australia?",
    options: [
      { id: "a", text: "Sydney" },
      { id: "b", text: "Melbourne" },
      { id: "c", text: "Brisbane" },
      { id: "d", text: "Canberra" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q27",
    text: "How many sides does a hexagon have?",
    options: [
      { id: "a", text: "5" },
      { id: "b", text: "6" },
      { id: "c", text: "7" },
      { id: "d", text: "8" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q28",
    text: "What is the chemical symbol for gold?",
    options: [
      { id: "a", text: "Go" },
      { id: "b", text: "Gd" },
      { id: "c", text: "Au" },
      { id: "d", text: "Ag" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q29",
    text: "Which planet is known as the Red Planet?",
    options: [
      { id: "a", text: "Venus" },
      { id: "b", text: "Jupiter" },
      { id: "c", text: "Saturn" },
      { id: "d", text: "Mars" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q30",
    text: "What is the fastest land animal?",
    options: [
      { id: "a", text: "Lion" },
      { id: "b", text: "Cheetah" },
      { id: "c", text: "Greyhound" },
      { id: "d", text: "Ostrich" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q31",
    text: "In what country was the game of chess invented?",
    options: [
      { id: "a", text: "China" },
      { id: "b", text: "Persia" },
      { id: "c", text: "India" },
      { id: "d", text: "Greece" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q32",
    text: "How many bones are in the adult human body?",
    options: [
      { id: "a", text: "196" },
      { id: "b", text: "206" },
      { id: "c", text: "216" },
      { id: "d", text: "226" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 20,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q33",
    text: "What is the longest river in the world?",
    options: [
      { id: "a", text: "Amazon" },
      { id: "b", text: "Mississippi" },
      { id: "c", text: "Yangtze" },
      { id: "d", text: "Nile" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q34",
    text: "What does 'HTTP' stand for?",
    options: [
      { id: "a", text: "HyperText Transfer Protocol" },
      { id: "b", text: "High-speed Text Transmission Platform" },
      { id: "c", text: "HyperText Transport Process" },
      { id: "d", text: "Home Terminal Transfer Protocol" },
    ],
    correctOptionId: "a",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q35",
    text: "Which element has atomic number 1?",
    options: [
      { id: "a", text: "Helium" },
      { id: "b", text: "Oxygen" },
      { id: "c", text: "Hydrogen" },
      { id: "d", text: "Carbon" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q36",
    text: "What is the tallest mountain in the world?",
    options: [
      { id: "a", text: "K2" },
      { id: "b", text: "Kangchenjunga" },
      { id: "c", text: "Mount Kilimanjaro" },
      { id: "d", text: "Mount Everest" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q37",
    text: "How many continents are there on Earth?",
    options: [
      { id: "a", text: "5" },
      { id: "b", text: "6" },
      { id: "c", text: "7" },
      { id: "d", text: "8" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q38",
    text: "What language has the most native speakers in the world?",
    options: [
      { id: "a", text: "English" },
      { id: "b", text: "Spanish" },
      { id: "c", text: "Mandarin Chinese" },
      { id: "d", text: "Hindi" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q39",
    text: "What is the square root of 144?",
    options: [
      { id: "a", text: "11" },
      { id: "b", text: "12" },
      { id: "c", text: "13" },
      { id: "d", text: "14" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q40",
    text: "How many players are on a standard basketball team on the court?",
    options: [
      { id: "a", text: "4" },
      { id: "b", text: "5" },
      { id: "c", text: "6" },
      { id: "d", text: "7" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },

  // ── FOOD & PARTY ────────────────────────────────────────────────────────────
  {
    id: "q41",
    text: "Which country invented pizza?",
    options: [
      { id: "a", text: "USA" },
      { id: "b", text: "Greece" },
      { id: "c", text: "Italy" },
      { id: "d", text: "France" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q42",
    text: "What is the main ingredient in guacamole?",
    options: [
      { id: "a", text: "Tomato" },
      { id: "b", text: "Avocado" },
      { id: "c", text: "Lime" },
      { id: "d", text: "Jalapeño" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q43",
    text: "What cocktail is made with vodka, tomato juice, and Worcestershire sauce?",
    options: [
      { id: "a", text: "Cosmopolitan" },
      { id: "b", text: "Bloody Mary" },
      { id: "c", text: "Screwdriver" },
      { id: "d", text: "Moscow Mule" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q44",
    text: "What is the most popular pizza topping in the USA?",
    options: [
      { id: "a", text: "Sausage" },
      { id: "b", text: "Mushrooms" },
      { id: "c", text: "Pepperoni" },
      { id: "d", text: "Extra cheese" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },

  // ── MUSIC HISTORY ───────────────────────────────────────────────────────────
  {
    id: "q45",
    text: "Which band wrote 'Bohemian Rhapsody'?",
    options: [
      { id: "a", text: "The Beatles" },
      { id: "b", text: "Led Zeppelin" },
      { id: "c", text: "Queen" },
      { id: "d", text: "The Rolling Stones" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q46",
    text: "Michael Jackson's 'Thriller' was released in which year?",
    options: [
      { id: "a", text: "1980" },
      { id: "b", text: "1982" },
      { id: "c", text: "1984" },
      { id: "d", text: "1986" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q47",
    text: "What genre of music originated in Jamaica in the 1960s?",
    options: [
      { id: "a", text: "Salsa" },
      { id: "b", text: "Reggae" },
      { id: "c", text: "Soca" },
      { id: "d", text: "Calypso" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q48",
    text: "Which artist is known as the 'Queen of Pop'?",
    options: [
      { id: "a", text: "Lady Gaga" },
      { id: "b", text: "Britney Spears" },
      { id: "c", text: "Whitney Houston" },
      { id: "d", text: "Madonna" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q49",
    text: "What does 'EDM' stand for?",
    options: [
      { id: "a", text: "Electronic Digital Music" },
      { id: "b", text: "Electronic Dance Music" },
      { id: "c", text: "Enhanced Digital Mixing" },
      { id: "d", text: "Electric Drive Music" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q50",
    text: "Which city is known as the birthplace of Jazz?",
    options: [
      { id: "a", text: "Chicago" },
      { id: "b", text: "New York" },
      { id: "c", text: "New Orleans" },
      { id: "d", text: "Memphis" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q51",
    text: "Kendrick Lamar is from which city?",
    options: [
      { id: "a", text: "Oakland" },
      { id: "b", text: "Compton" },
      { id: "c", text: "Atlanta" },
      { id: "d", text: "Chicago" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q52",
    text: "What is the term for a DJ using two copies of the same record to extend a break?",
    options: [
      { id: "a", text: "Beat matching" },
      { id: "b", text: "Scratching" },
      { id: "c", text: "The Merry-Go-Round technique" },
      { id: "d", text: "Looping" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "hard",
  },
  {
    id: "q53",
    text: "Which producer is known as 'The Neptunes'?",
    options: [
      { id: "a", text: "Timbaland" },
      { id: "b", text: "Kanye West" },
      { id: "c", text: "Pharrell Williams & Chad Hugo" },
      { id: "d", text: "Dr. Dre" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "hard",
  },
  {
    id: "q54",
    text: "What is the tempo range typically associated with Techno music?",
    options: [
      { id: "a", text: "60–80 BPM" },
      { id: "b", text: "90–110 BPM" },
      { id: "c", text: "130–160 BPM" },
      { id: "d", text: "180–200 BPM" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q55",
    text: "What Grammy category did Beyoncé win the most times in a single year?",
    options: [
      { id: "a", text: "Album of the Year" },
      { id: "b", text: "Record of the Year" },
      { id: "c", text: "Most Grammys in one night (4)" },
      { id: "d", text: "Best Pop Album" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "pop culture",
    difficulty: "hard",
  },

  // ── BATCH 3 — Party / Food / Random ─────────────────────────────────────────
  {
    id: "q56",
    text: "How many sides does a hexagon have?",
    options: [
      { id: "a", text: "5" },
      { id: "b", text: "6" },
      { id: "c", text: "7" },
      { id: "d", text: "8" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q57",
    text: "What is the most spoken language in the world by total speakers?",
    options: [
      { id: "a", text: "Spanish" },
      { id: "b", text: "Mandarin Chinese" },
      { id: "c", text: "English" },
      { id: "d", text: "Hindi" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q58",
    text: "Which planet has the most moons?",
    options: [
      { id: "a", text: "Jupiter" },
      { id: "b", text: "Saturn" },
      { id: "c", text: "Uranus" },
      { id: "d", text: "Neptune" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q59",
    text: "What country invented pizza?",
    options: [
      { id: "a", text: "Greece" },
      { id: "b", text: "France" },
      { id: "c", text: "Italy" },
      { id: "d", text: "Spain" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "food",
    difficulty: "easy",
  },
  {
    id: "q60",
    text: "How many calories are in one gram of alcohol?",
    options: [
      { id: "a", text: "4" },
      { id: "b", text: "5" },
      { id: "c", text: "7" },
      { id: "d", text: "9" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 20,
    category: "food",
    difficulty: "hard",
  },
  {
    id: "q61",
    text: "In which city was hip-hop born?",
    options: [
      { id: "a", text: "Los Angeles" },
      { id: "b", text: "Atlanta" },
      { id: "c", text: "New York City" },
      { id: "d", text: "Chicago" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q62",
    text: "What does EDM stand for?",
    options: [
      { id: "a", text: "Electronic Dance Music" },
      { id: "b", text: "Every DJ Mixes" },
      { id: "c", text: "Electric Digital Melody" },
      { id: "d", text: "Enhanced Dance Mode" },
    ],
    correctOptionId: "a",
    timeLimitSeconds: 10,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q63",
    text: "Which artist holds the record for most Grammy wins of all time?",
    options: [
      { id: "a", text: "Adele" },
      { id: "b", text: "Beyoncé" },
      { id: "c", text: "Taylor Swift" },
      { id: "d", text: "Michael Jackson" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q64",
    text: "What is the name of Taylor Swift's 2023 global tour?",
    options: [
      { id: "a", text: "Reputation Tour" },
      { id: "b", text: "Lover Fest" },
      { id: "c", text: "The Eras Tour" },
      { id: "d", text: "Midnights World Tour" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q65",
    text: "Which streaming service created 'Stranger Things'?",
    options: [
      { id: "a", text: "HBO Max" },
      { id: "b", text: "Hulu" },
      { id: "c", text: "Amazon Prime" },
      { id: "d", text: "Netflix" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 10,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q66",
    text: "What is the fastest land animal?",
    options: [
      { id: "a", text: "Lion" },
      { id: "b", text: "Cheetah" },
      { id: "c", text: "Pronghorn Antelope" },
      { id: "d", text: "Greyhound" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q67",
    text: "How many strings does a standard guitar have?",
    options: [
      { id: "a", text: "4" },
      { id: "b", text: "5" },
      { id: "c", text: "6" },
      { id: "d", text: "7" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 10,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q68",
    text: "What year was the first iPhone released?",
    options: [
      { id: "a", text: "2005" },
      { id: "b", text: "2006" },
      { id: "c", text: "2007" },
      { id: "d", text: "2008" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q69",
    text: "Which country consumes the most sushi per capita?",
    options: [
      { id: "a", text: "United States" },
      { id: "b", text: "Japan" },
      { id: "c", text: "South Korea" },
      { id: "d", text: "China" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "food",
    difficulty: "easy",
  },
  {
    id: "q70",
    text: "What cocktail is made with vodka, lime juice, and ginger beer?",
    options: [
      { id: "a", text: "Mojito" },
      { id: "b", text: "Moscow Mule" },
      { id: "c", text: "Margarita" },
      { id: "d", text: "Dark and Stormy" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "food",
    difficulty: "medium",
  },
  {
    id: "q71",
    text: "How many players are on a basketball team on the court at one time?",
    options: [
      { id: "a", text: "4" },
      { id: "b", text: "5" },
      { id: "c", text: "6" },
      { id: "d", text: "7" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q72",
    text: "Which element has the chemical symbol 'Au'?",
    options: [
      { id: "a", text: "Silver" },
      { id: "b", text: "Aluminum" },
      { id: "c", text: "Gold" },
      { id: "d", text: "Copper" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q73",
    text: "Who sang 'Bad Guy' released in 2019?",
    options: [
      { id: "a", text: "Ariana Grande" },
      { id: "b", text: "Dua Lipa" },
      { id: "c", text: "Billie Eilish" },
      { id: "d", text: "Lizzo" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q74",
    text: "What is the longest river in the world?",
    options: [
      { id: "a", text: "Amazon" },
      { id: "b", text: "Mississippi" },
      { id: "c", text: "Yangtze" },
      { id: "d", text: "Nile" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
  {
    id: "q75",
    text: "Which decade did the 'rave' party scene explode in the UK?",
    options: [
      { id: "a", text: "1970s" },
      { id: "b", text: "1980s" },
      { id: "c", text: "1990s" },
      { id: "d", text: "2000s" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "medium",
  },
  {
    id: "q76",
    text: "What colour is the Spotify logo?",
    options: [
      { id: "a", text: "Blue" },
      { id: "b", text: "Green" },
      { id: "c", text: "Orange" },
      { id: "d", text: "Purple" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "pop culture",
    difficulty: "easy",
  },
  {
    id: "q77",
    text: "Which country hosted the 2022 FIFA World Cup?",
    options: [
      { id: "a", text: "Saudi Arabia" },
      { id: "b", text: "UAE" },
      { id: "c", text: "Qatar" },
      { id: "d", text: "Bahrain" },
    ],
    correctOptionId: "c",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "easy",
  },
  {
    id: "q78",
    text: "What is the main ingredient in guacamole?",
    options: [
      { id: "a", text: "Mango" },
      { id: "b", text: "Avocado" },
      { id: "c", text: "Lime" },
      { id: "d", text: "Tomato" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 10,
    category: "food",
    difficulty: "easy",
  },
  {
    id: "q79",
    text: "DJ Khaled is known for shouting what catchphrase?",
    options: [
      { id: "a", text: "We the best!" },
      { id: "b", text: "Let's go!" },
      { id: "c", text: "Turn it up!" },
      { id: "d", text: "Another one!" },
    ],
    correctOptionId: "d",
    timeLimitSeconds: 15,
    category: "music",
    difficulty: "easy",
  },
  {
    id: "q80",
    text: "How many bones are in the adult human body?",
    options: [
      { id: "a", text: "186" },
      { id: "b", text: "206" },
      { id: "c", text: "216" },
      { id: "d", text: "226" },
    ],
    correctOptionId: "b",
    timeLimitSeconds: 15,
    category: "general",
    difficulty: "medium",
  },
];
